### PR TITLE
zod schemas

### PR DESCRIPTION
**Pull request by bergarces**
Minor changes
- Removed question that allows to give the adapter class a different name. It is now deterministic based on protocol and product. Necessary for deterministic file paths.
- Renamed adapters that were not following that convention.
- Moved dotenv as a dev dependency for jest. CLI uses node 20 in-built system.

Major changes
- Added export of `WriteActionInputs` at adapter level, which is an object with the zod schemas for the input of each action. This is needed to generate types for the SDK.
- Added export of `GetTransactionParamsSchema` at adapter level, which is a zod schema for all the possible parameters that the `getTransactionParams` method can receive. This is also exported so it can be used in the lambdas. (This whole object could be automatically generated based on the write action inputs, which is a TODO).
- Automatically centralise all `WriteActionInputs` and `GetTransactionParamsSchema` into `src/adapters/index.ts` so they can easily be updated and exported. They get updated with `npm run build-types`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -89,7 +90,6 @@
     "typescript": "^5.1.6"
   },
   "dependencies": {
-    "dotenv": "^16.3.1",
```
Comment (bergarces) [inline package.json:92]: It's not required as dev dependency, but still needed for tests.

File: src/adapters/index.ts [hunk 2]
```diff
@@ -302,3 +315,18 @@ export const supportedProtocols: Record<
...
+}
+
+export const GetTransactionParamsInputSchema = z.union([
+  AaveV3ATokenGetTxParamsInput,
+  CompoundV2BorrowMarketGetTxParamsInput,
+  CompoundV2SupplyMarketGetTxParamsInput,
+])
+export type GetTransactionParamsInput = z.infer<
+  typeof GetTransactionParamsInputSchema
+>
```
Comment (bergarces) [inline src/adapters/index.ts:332]: This code is automatically generated when calling `npm run build-types`.

It'll be based on the exports from the adapter.

File: src/scripts/newAdapterCommand.ts [hunk 3]
```diff
@@ -186,49 +186,6 @@ export function newAdapterCommand(
...
-                  productId,
-                  input,
-                )
-              ) {
-                return 'An adapter with that name already exists'
-              }
-
-              return true
-            },
-          },
```
Comment (bergarces) [inline src/scripts/newAdapterCommand.ts:231]: Removed question and setting adapter class and file name based on `protocolId` and `productId`.

We are already checking on `productId` for uniqueness.

File: src/scripts/buildTypes.ts [hunk 4]
```diff
@@ -64,3 +73,161 @@ async function buildContractTypesForFolder(contractsDirectory: string) {
...
+  productId: string
+}) {
+  const adaptersFile = path.resolve('./src/adapters/index.ts')
+  const contents = fs.readFileSync(adaptersFile, 'utf-8')
+  const ast = parse(contents, {
+    parser: require('recast/parsers/typescript'),
+  })
+
+  visit(ast, {
+    visitImportDeclaration(path) {
```
Comment (bergarces) [inline src/scripts/buildTypes.ts:124]: This is to add the new elements to the import declaration.

File: src/defiProvider.ts [hunk 5]
```diff
@@ -543,9 +540,9 @@ export class DefiProvider {
     })
   }
 
-  getSupport(input: {
-    filterChainIds: Chain[] | undefined
-    filterProtocolIds: Protocol[] | undefined
+  getSupport(input?: {
+    filterChainIds?: Chain[] | undefined
+    filterProtocolIds?: Protocol[] | undefined
```
Comment (bergarces) [inline src/defiProvider.ts:545]: Make all these filters optional.

File: src/index.ts [hunk 6]
```diff
@@ -7,4 +7,5 @@ export type {
   DefiPositionResponse,
   DefiProfitsResponse,
 } from './types/response'
-export type { GetTransactionParamsInput } from './types/getTransactionParamsInput'
+export type { GetTransactionParamsInput } from './adapters'
+export { GetTransactionParamsInputSchema } from './adapters'
```
Comment (bergarces) [inline src/index.ts:11]: This exported schema will be used with the lambdas to add validation.

File: src/index.ts [hunk 7]
```diff
@@ -4,7 +4,10 @@ export { TimePeriod } from './core/constants/timePeriod'
...
 export { PositionType } from './types/adapter'
 export type {
+  AdapterResponse,
   DefiPositionResponse,
   DefiProfitsResponse,
 } from './types/response'
-export type { GetTransactionParamsInput } from './types/getTransactionParamsInput'
+export { WriteActions } from './types/writeActions'
+export type { GetTransactionParams } from './adapters'
+export { GetTransactionParamsSchema, WriteActionInputs } from './adapters'
```
Comment (bergarces) [inline src/index.ts:13]: Added a few more exports needed for lambdas and sdk packages.

File: src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts [hunk 8]
```diff
@@ -84,7 +84,7 @@ export class MendiFinanceSupplyAdapter
    * Implementation Steps:
    * 1. Implement logic for handling predefined actions (e.g., Supply, Withdraw). Consider the examples provided as a starting point.
    * 2. For new actions (e.g., Stake, Flash Loan), first extend the `WriteActions` object to include these new actions.
-   * 3. Update `GetTransactionParamsInput` type to include the parameters required for any new actions you add.
+   * 3. Export WriteActionInputs, GetTransactionParamsSchema and GetTransactionParams from this file.
```
Comment (jpsains) [inline src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts:87]: Does this need to be updated in the templates too?

File: src/adapters/index.ts [hunk 9]
```diff
@@ -316,3 +329,16 @@ export const supportedProtocols: Record<
...
+  CompoundV2BorrowMarketWriteActionInputs,
+  CompoundV2SupplyMarketWriteActionInputs,
+}
+
+export const GetTransactionParamsSchema = z.union([
+  AaveV3ATokenGetTransactionParamsSchema,
+  CompoundV2BorrowMarketGetTransactionParamsSchema,
+  CompoundV2SupplyMarketGetTransactionParamsSchema,
+])
+export type GetTransactionParams = z.infer<typeof GetTransactionParamsSchema>
```
Comment (jpsains) [inline src/adapters/index.ts:344]: just to align: this is updated automatically when a GetTransactionParams type is exported from an adapter file?

File: src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts (hunk 8 above)
Comment (jpsains) [reply src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts:87]: if so maybe we can make it one string thats reused in all templates e.g. const writeAdapterComment = '.....'

File: src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts [hunk 10]
```diff
@@ -52,13 +50,7 @@ export class AaveV3ATokenPoolAdapter extends AaveBasePoolAdapter {
     return reserveData.liquidityRate
   }
```
Comment (jpsains) [inline src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts:51]: the templates import get transaction params from `import { GetTransactionParams } from '../../..'`

Should we keep consistent?

**Review by jpsains — APPROVED** [review]

File: src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts (hunk 10 above)
Comment (bergarces) [reply src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts:51]: Yes. There are advantages of doing it that way. I changed it for the already implemented ones and forgot to revert.

File: src/adapters/index.ts (hunk 9 above)
Comment (bergarces) [reply src/adapters/index.ts:344]: We need to run `npm run build-types`

File: src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts (hunk 8 above)
Comment (bergarces) [reply src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts:87]: It's been updated there too.

Comment (bergarces) [reply src/adapters/mendi-finance/products/supply/mendiFinanceSupplyAdapter.ts:87]: What we need is now a better explanation of what it is.

But I'm going to create a ticket to automate the creation of `GetTransactionParamsSchema` object and type, then we can add the explanation there, which will be much simpler.

File: src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts (hunk 10 above)
Comment (bergarces) [reply src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts:51]: Actually. I'll have to change the templates, as the name clashes.

**Pull request merged**